# HOUSE PRICING

Pricing is one of the most searched areas in many industries, since it allows companies to get ahead of the game of sales. Concretely, finding out which are the house features that influence on the price of a house will give the key to predict the price of houses in the market.

Ask a home buyer to describe their dream house, and they probably won't begin with the height of the basement ceiling or the proximity to an east-west railroad. But this playground competition's dataset proves that much more influences price negotiations than the number of bedrooms or a white-picket fence.

With 79 explanatory variables describing (almost) every aspect of residential homes in Ames, Iowa, this competition challenges to predict the final price of each home.

This version of the code containt the procedure of creating the functions. It also described why some techniques for correlation do not work for this dataset.

# Data cleaning

Let's start improting some libraries for EDA

In [1]:
import pandas as pd
import numpy as np

import seaborn as sns
from scipy import stats # Correlation analysis
import scipy.stats as ss

Upload training dataset and explore the data:

In [2]:
df_train = pd.read_csv("train.csv")
df_train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
5         Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
7         Lvl    AllPub  ...        0    NaN    NaN        Shed     350   
8         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
9         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

In [3]:
df_train.shape

(1460, 81)

Resize the maximum number of columns and rows that can be displayed to view all the columns:

In [4]:
pd.set_option('display.max_columns', 81)
pd.set_option('display.max_rows', 81)
df_train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
5         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
6         Lvl    AllPub    Inside       Gtl      Somerst       Norm   
7         Lvl    AllPub    Corner       Gtl       NWAmes       PosN   
8         Lvl    AllPub    Inside       Gtl      OldTown     Artery   
9         Lvl    AllPub    Corner       Gtl      BrkSide     Artery   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   
5       Norm     1Fam     1.5Fin            5            5       1993   
6       Norm     1Fam     1Story            8            5       2004   
7       Norm     1Fam     2Story            7            6       1973   
8       Norm     1Fam     1.5Fin            7            5       1931   
9     Artery   2fmCon     1.5Unf            5            6       1939   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
5          1995     Gable  CompShg     VinylSd     VinylSd       None   
6          2005     Gable  CompShg     VinylSd     VinylSd      Stone   
7          1973     Gable  CompShg     HdBoard     HdBoard      Stone   
8          1950     Gable  CompShg     BrkFace     Wd Shng       None   
9          1950     Gable  CompShg     MetalSd     MetalSd       None   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   
5         0.0        TA        TA       Wood       Gd       TA           No   
6       186.0        Gd        TA      PConc       Ex       TA           Av   
7       240.0        TA        TA     CBlock       Gd       TA           Mn   
8         0.0        TA        TA     BrkTil       TA       TA           No   
9       

In [5]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [6]:
# Describe numeric data
df_train.describe().T

count           mean           std      min        25%  \
Id             1460.0     730.500000    421.610009      1.0     365.75   
MSSubClass     1460.0      56.897260     42.300571     20.0      20.00   
LotFrontage    1201.0      70.049958     24.284752     21.0      59.00   
LotArea        1460.0   10516.828082   9981.264932   1300.0    7553.50   
OverallQual    1460.0       6.099315      1.382997      1.0       5.00   
OverallCond    1460.0       5.575342      1.112799      1.0       5.00   
YearBuilt      1460.0    1971.267808     30.202904   1872.0    1954.00   
YearRemodAdd   1460.0    1984.865753     20.645407   1950.0    1967.00   
MasVnrArea     1452.0     103.685262    181.066207      0.0       0.00   
BsmtFinSF1     1460.0     443.639726    456.098091      0.0       0.00   
BsmtFinSF2     1460.0      46.549315    161.319273      0.0       0.00   
BsmtUnfSF      1460.0     567.240411    441.866955      0.0     223.00   
TotalBsmtSF    1460.0    1057.429452    438.705324      0.0     795.75   
1stFlrSF       1460.0    1162.626712    386.587738    334.0     882.00   
2ndFlrSF       1460.0     346.992466    436.528436      0.0       0.00   
LowQualFinSF   1460.0       5.844521     48.623081      0.0       0.00   
GrLivArea      1460.0    1515.463699    525.480383    334.0    1129.50   
BsmtFullBath   1460.0       0.425342      0.518911      0.0       0.00   
BsmtHalfBath   1460.0       0.057534      0.238753      0.0       0.00   
FullBath       1460.0       1.565068      0.550916      0.0       1.00   
HalfBath       1460.0       0.382877      0.502885      0.0       0.00   
BedroomAbvGr   1460.0       2.866438      0.815778      0.0       2.00   
KitchenAbvGr   1460.0       1.046575      0.220338      0.0       1.00   
TotRmsAbvGrd   1460.0       6.517808      1.625393      2.0       5.00   
Fireplaces     1460.0       0.613014      0.644666      0.0       0.00   
GarageYrBlt    1379.0    1978.506164     24.689725   1900.0    1961.00   
GarageCars     1460.0       1.767123      0.747315      0.0       1.00   
GarageArea     1460.0     472.980137    213.804841      0.0     334.50   
WoodDeckSF     1460.0      94.244521    125.338794      0.0       0.00   
OpenPorchSF    1460.0      46.660274     66.256028      0.0       0.00   
EnclosedPorch  1460.0      21.954110     61.119149      0.0       0.00   
3SsnPorch      1460.0       3.409589     29.317331      0.0       0.00   
ScreenPorch    1460.0      15.060959     55.757415      0.0       0.00   
PoolArea       1460.0       2.758904     40.177307      0.0       0.00   
MiscVal        1460.0      43.489041    496.123024      0.0       0.00   
MoSold         1460.0       6.321918      2.703626      1.0       5.00   
YrSold         1460.0    2007.815753      1.328095   2006.0    2007.00   
SalePrice      1460.0  180921.195890  79442.502883  34900.0  129975.00   

                    50%        75%       max  
Id                730.5    1095.25    1460.0  
MSSubClass         50.0      70.00     190.0  
LotFrontage        69.0      80.00     313.0  
LotArea          9478.5   11601.50  215245.0  
OverallQual         6.0       7.00      10.0  
OverallCond         5.0       6.00       9.0  
YearBuilt        1973.0    2000.00    2010.0  
YearRemodAdd     1994.0    2004.00    2010.0  
MasVnrArea          0.0     166.00    1600.0  
BsmtFinSF1        383.5     712.25    5644.0  
BsmtFinSF2          0.0       0.00    1474.0  
BsmtUnfSF         477.5     808.00    2336.0  
TotalBsmtSF       991.5    1298.25    6110.0  
1stFlrSF         1087.0    1391.25    4692.0  
2ndFlrSF            0.0     728.00    2065.0  
LowQualFinSF        0.0       0.00     572.0  
GrLivArea        1464.0    1776.75    5642.0  
BsmtFullBath        0.0       1.00       3.0  
BsmtHalfBath        0.0       0.00       2.0  
FullBath            2.0       2.00       3.0  
HalfBath            0.0       1.00       2.0  
BedroomAbvGr        3.0       3.00       8.0  
KitchenAbvGr        1.0       1.00       3.0  
T

In [7]:
# Describe object data
df_train.describe(include='O').T

count unique      top  freq
MSZoning       1460      5       RL  1151
Street         1460      2     Pave  1454
Alley            91      2     Grvl    50
LotShape       1460      4      Reg   925
LandContour    1460      4      Lvl  1311
Utilities      1460      2   AllPub  1459
LotConfig      1460      5   Inside  1052
LandSlope      1460      3      Gtl  1382
Neighborhood   1460     25    NAmes   225
Condition1     1460      9     Norm  1260
Condition2     1460      8     Norm  1445
BldgType       1460      5     1Fam  1220
HouseStyle     1460      8   1Story   726
RoofStyle      1460      6    Gable  1141
RoofMatl       1460      8  CompShg  1434
Exterior1st    1460     15  VinylSd   515
Exterior2nd    1460     16  VinylSd   504
MasVnrType     1452      4     None   864
ExterQual      1460      4       TA   906
ExterCond      1460      5       TA  1282
Foundation     1460      6    PConc   647
BsmtQual       1423      4       TA   649
BsmtCond       1423      4       TA  1311
BsmtExposure   1422      4       No   953
BsmtFinType1   1423      6      Unf   430
BsmtFinType2   1422      6      Unf  1256
Heating        1460      6     GasA  1428
HeatingQC      1460      5       Ex   741
CentralAir     1460      2        Y  1365
Electrical     1459      5    SBrkr  1334
KitchenQual    1460      4       TA   735
Functional     1460      7      Typ  1360
FireplaceQu     770      5       Gd   380
GarageType     1379      6   Attchd   870
GarageFinish   1379      3      Unf   605
GarageQual     1379      5       TA  1311
GarageCond     1379      5       TA  1326
PavedDrive     1460      3        Y  1340
PoolQC            7      3       Gd     3
Fence           281      4    MnPrv   157
MiscFeature      54      4     Shed    49
SaleType       1460      9       WD  1267
SaleCondition  1460      6   Normal  1198

## Missing values

We saw many NaN values on the dataframe overview. Let's calculate the percentage of the missing values in each attribute:

In [8]:
df_train.isnull().mean()*100

Id                0.000000
MSSubClass        0.000000
MSZoning          0.000000
LotFrontage      17.739726
LotArea           0.000000
Street            0.000000
Alley            93.767123
LotShape          0.000000
LandContour       0.000000
Utilities         0.000000
LotConfig         0.000000
LandSlope         0.000000
Neighborhood      0.000000
Condition1        0.000000
Condition2        0.000000
BldgType          0.000000
HouseStyle        0.000000
OverallQual       0.000000
OverallCond       0.000000
YearBuilt         0.000000
YearRemodAdd      0.000000
RoofStyle         0.000000
RoofMatl          0.000000
Exterior1st       0.000000
Exterior2nd       0.000000
MasVnrType        0.547945
MasVnrArea        0.547945
ExterQual         0.000000
ExterCond         0.000000
Foundation        0.000000
BsmtQual          2.534247
BsmtCond          2.534247
BsmtExposure      2.602740
BsmtFinType1      2.534247
BsmtFinSF1        0.000000
BsmtFinType2      2.602740
BsmtFinSF2        0.000000
B

In [9]:
# Collect the features that have missing values
missing_values = []
for nan in df_train.columns:
    threshold = df_train[nan].isnull().mean()
    if threshold > 0:
        missing_values.append(nan)
        #print(threshold)
        
# Which can be written as follows:
missing_values = [nan for nan in df_train.columns if df_train[nan].isnull().mean() > 0]

In [10]:
# Unite features and values and create dataframe from array
nan_values = df_train[missing_values].isnull().mean()
nan_values = pd.DataFrame(nan_values.reset_index())#.rename(columns = {'index':'Feature', 0:'Null%'})
# Add column names
nan_values.columns = ['Feature', 'Null%']

# Show data types
data_type = [df_train[i].dtype for i in nan_values['Feature']]
nan_values['Data_type'] = data_type

# Transform mean into a percentage
nan_values['Null%'] = (nan_values['Null%']*100).round(decimals = 2)
# Sort values
nan_values.sort_values('Null%', ascending=False, inplace=True)
# Add a color gradient
color = sns.light_palette("pink", as_cmap=True)
nan_values.style.background_gradient(cmap=color)

The top 4 features could be dropped, given that they are missing more than 50% of the values.

In [11]:
dataset=df_train

In [12]:
thres = 50

"""# Drop columns which % of null values is higher than "thres"
for i,row in nan_values.iterrows():
    if row['Null%'] > thres:
        new_dataset = dataset.drop(columns=row['Feature'], inplace=True)
        """
# It can also be written as follows and drop the features afterwards:

# Collect columns which % of null values is higher than "thres"
drop_cols = [row['Feature'] for i,row in nan_values.iterrows() if row['Null%'] > thres]
drop_cols

['PoolQC', 'MiscFeature', 'Alley', 'Fence']

In [13]:
new_dataset = dataset.drop(columns=drop_cols)
new_dataset

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
0        1          60       RL         65.0     8450   Pave      Reg   
1        2          20       RL         80.0     9600   Pave      Reg   
2        3          60       RL         68.0    11250   Pave      IR1   
3        4          70       RL         60.0     9550   Pave      IR1   
4        5          60       RL         84.0    14260   Pave      IR1   
...    ...         ...      ...          ...      ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      Reg   
1456  1457          20       RL         85.0    13175   Pave      Reg   
1457  1458          70       RL         66.0     9042   Pave      Reg   
1458  1459          20       RL         68.0     9717   Pave      Reg   
1459  1460          20       RL         75.0     9937   Pave      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   
..

Now we need to deal with the other missing values. We will substitute them with the mode of the feature in the case of numerical values, and with the frecuency for the object ones. \
We will have to be alert on the FireplaceQu	and LotFrontage results, since they are the ones that are missing a higher number of values.

In [14]:
# Collect columns which % of null values is equal or lower than "thres"
fill_cols = [row['Feature'] for i,row in nan_values.iterrows() if row['Null%'] <= thres]

# Create dataframe and merge to obtain their data types
df_fill_cols = pd.DataFrame()
df_fill_cols['Feature'] = fill_cols
df_fill_cols = df_fill_cols.merge(nan_values).drop(columns='Null%')
df_fill_cols

Feature Data_type
0    FireplaceQu    object
1    LotFrontage   float64
2     GarageType    object
3    GarageYrBlt   float64
4   GarageFinish    object
5     GarageQual    object
6     GarageCond    object
7   BsmtExposure    object
8   BsmtFinType2    object
9   BsmtFinType1    object
10      BsmtCond    object
11      BsmtQual    object
12    MasVnrArea   float64
13    MasVnrType    object
14    Electrical    object

In [15]:
# if object -> frequancy; if number -> mean

for i,row in df_fill_cols.iterrows():
    #print(row)
    feature = row['Feature']
    if row['Data_type'] == 'object':
        #print('object')
        freq_max = new_dataset[feature].value_counts().idxmax()
        new_dataset[feature].replace(np.nan, freq_max, inplace=True)
    else:
        #print('number')
        #avg_val = new_dataset[feature].astype("float").mean(axis=0)
        #new_dataset[feature].replace(np.nan, avg_val, inplace=True)
        mode_val = new_dataset[feature].mode()[0]
        new_dataset[feature].replace(np.nan, mode_val, inplace=True)
new_dataset

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
0        1          60       RL         65.0     8450   Pave      Reg   
1        2          20       RL         80.0     9600   Pave      Reg   
2        3          60       RL         68.0    11250   Pave      IR1   
3        4          70       RL         60.0     9550   Pave      IR1   
4        5          60       RL         84.0    14260   Pave      IR1   
...    ...         ...      ...          ...      ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      Reg   
1456  1457          20       RL         85.0    13175   Pave      Reg   
1457  1458          70       RL         66.0     9042   Pave      Reg   
1458  1459          20       RL         68.0     9717   Pave      Reg   
1459  1460          20       RL         75.0     9937   Pave      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   
..

In [16]:
target ='SalePrice'
# Drop rows that are nan on the target column:
new_dataset.dropna(subset=[target], axis=0, inplace=True)
new_dataset.reset_index(drop=True, inplace=True)

Let's put it in a function:

In [17]:
def check_fill_nulls(dataset, thres, target, opt_fill=0):
    
    """
    This funtion checks the % of nulls for each feature in a given dataframe.
    It returns:
        - A table with a color gradient that is more intens according to the % of missing values.
        - A list of which are these dropped features.
        - A new dataframe that excludes those features from which more than a certain given % (thres) 
          of their values are missing.
    The new dataframe will exclude those rows where the target variable is nan, and will also fill in the 
    nulls values (numeric for mean, object for max frequency) if opt_fill=1.
    """
    
    # Create a table extracting the NaN values from the dataset
    missing_values = [nan for nan in dataset.columns if dataset[nan].isnull().mean() > 0]
    
    # Unite features and values and create dataframe from array
    nan_values = dataset[missing_values].isnull().mean()
    nan_values = pd.DataFrame(nan_values.reset_index())
    
    # Add column names
    nan_values.columns = ['Feature', 'Null%']
    
    # Show data types
    data_type = [df_train[i].dtype for i in nan_values['Feature']]
    nan_values['Data_type'] = data_type
    
    # Transform mean into a percentage
    nan_values['Null%'] = (nan_values['Null%']*100).round(decimals = 2)
    
    # Sort values and add a color gradient
    nan_values.sort_values('Null%', ascending=False, inplace=True)
    color = sns.light_palette("pink", as_cmap=True)
    colored_df = nan_values.style.background_gradient(cmap=color)
    
    # Collect columns which % of null values is higher than "thres"
    drop_cols = [row['Feature'] for i,row in nan_values.iterrows() if row['Null%'] > thres]
    
    # Drop collected features
    new_dataset = dataset.drop(columns=drop_cols)
    print("The following columns have been dropped:", drop_cols)
    
    # Drop rows that are nan on the target column:
    new_dataset.dropna(subset=[target], axis=0, inplace=True)
    new_dataset.reset_index(drop=True, inplace=True)
    
    # Filling the nulls
    # We will substitute them with the mean of the feature in the case of numerical values,
    # and with the frecuency for the object ones.
    if opt_fill == 1:
        
        # Collect columns which % of null values is equal or lower than "thres"
        fill_cols = [row['Feature'] for i,row in nan_values.iterrows() if row['Null%'] <= thres]

        # Create dataframe and merge to obtain their data types
        df_fill_cols = pd.DataFrame()
        df_fill_cols['Feature'] = fill_cols
        df_fill_cols = df_fill_cols.merge(nan_values).drop(columns='Null%')
        
        for i,row in df_fill_cols.iterrows():
        #print(row)
            feature = row['Feature']
            if row['Data_type'] == 'object':
                #print('object')
                freq_max = new_dataset[feature].value_counts().idxmax()
                new_dataset[feature].replace(np.nan, freq_max, inplace=True)
            else:
                #print('number')
                avg_val = new_dataset[feature].astype("float").mean(axis=0)
                new_dataset[feature].replace(np.nan, avg_val, inplace=True)
    
    
    return colored_df, drop_cols, new_dataset

In [18]:
ranking, dropped_cols, train = check_fill_nulls(df_train, thres=50, target='SalePrice', opt_fill=1)

The following columns have been dropped: ['PoolQC', 'MiscFeature', 'Alley', 'Fence']


In [19]:
train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
0        1          60       RL         65.0     8450   Pave      Reg   
1        2          20       RL         80.0     9600   Pave      Reg   
2        3          60       RL         68.0    11250   Pave      IR1   
3        4          70       RL         60.0     9550   Pave      IR1   
4        5          60       RL         84.0    14260   Pave      IR1   
...    ...         ...      ...          ...      ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      Reg   
1456  1457          20       RL         85.0    13175   Pave      Reg   
1457  1458          70       RL         66.0     9042   Pave      Reg   
1458  1459          20       RL         68.0     9717   Pave      Reg   
1459  1460          20       RL         75.0     9937   Pave      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   
..

In [20]:
train.isnull().mean()

Id               0.0
MSSubClass       0.0
MSZoning         0.0
LotFrontage      0.0
LotArea          0.0
Street           0.0
LotShape         0.0
LandContour      0.0
Utilities        0.0
LotConfig        0.0
LandSlope        0.0
Neighborhood     0.0
Condition1       0.0
Condition2       0.0
BldgType         0.0
HouseStyle       0.0
OverallQual      0.0
OverallCond      0.0
YearBuilt        0.0
YearRemodAdd     0.0
RoofStyle        0.0
RoofMatl         0.0
Exterior1st      0.0
Exterior2nd      0.0
MasVnrType       0.0
MasVnrArea       0.0
ExterQual        0.0
ExterCond        0.0
Foundation       0.0
BsmtQual         0.0
BsmtCond         0.0
BsmtExposure     0.0
BsmtFinType1     0.0
BsmtFinSF1       0.0
BsmtFinType2     0.0
BsmtFinSF2       0.0
BsmtUnfSF        0.0
TotalBsmtSF      0.0
Heating          0.0
HeatingQC        0.0
CentralAir       0.0
Electrical       0.0
1stFlrSF         0.0
2ndFlrSF         0.0
LowQualFinSF     0.0
GrLivArea        0.0
BsmtFullBath     0.0
BsmtHalfBath 

In [21]:
ranking

## Correlation

Let's explore now the correlation between the features:

The Pearson Correlation measures the linear dependence between two variables X and Y.

The resulting coefficient is a value between -1 and 1 inclusive, where:

- 1: Perfect positive linear correlation.
- 0: No linear correlation, the two variables most likely do not affect each other.
- -1: Perfect negative linear correlation.

In [22]:
train.corr()

Id  MSSubClass  LotFrontage   LotArea  OverallQual  \
Id             1.000000    0.011156    -0.009601 -0.033226    -0.028365   
MSSubClass     0.011156    1.000000    -0.357056 -0.139781     0.032628   
LotFrontage   -0.009601   -0.357056     1.000000  0.306795     0.234196   
LotArea       -0.033226   -0.139781     0.306795  1.000000     0.105806   
OverallQual   -0.028365    0.032628     0.234196  0.105806     1.000000   
OverallCond    0.012609   -0.059316    -0.052820 -0.005636    -0.091932   
YearBuilt     -0.012713    0.027850     0.117598  0.014228     0.572323   
YearRemodAdd  -0.021998    0.040581     0.082746  0.013788     0.550684   
MasVnrArea    -0.050199    0.022895     0.179283  0.103960     0.410238   
BsmtFinSF1    -0.005024   -0.069836     0.215828  0.214103     0.239666   
BsmtFinSF2    -0.005968   -0.065649     0.043340  0.111170    -0.059119   
BsmtUnfSF     -0.007940   -0.140759     0.122156 -0.002618     0.308159   
TotalBsmtSF   -0.015415   -0.238518     0.363358  0.260833     0.537808   
1stFlrSF       0.010496   -0.251758     0.414266  0.299475     0.476224   
2ndFlrSF       0.005590    0.307886     0.072483  0.050986     0.295493   
LowQualFinSF  -0.044230    0.046474     0.036849  0.004779    -0.030429   
GrLivArea      0.008273    0.074853     0.368392  0.263116     0.593007   
BsmtFullBath   0.002289    0.003491     0.091481  0.158155     0.111098   
BsmtHalfBath  -0.020155   -0.002333    -0.006419  0.048046    -0.040150   
FullBath       0.005587    0.131608     0.180424  0.126031     0.550600   
HalfBath       0.006784    0.177354     0.048258  0.014259     0.273458   
BedroomAbvGr   0.037719   -0.023438     0.237023  0.119690     0.101676   
KitchenAbvGr   0.002951    0.281721    -0.005805 -0.017784    -0.183882   
TotRmsAbvGrd   0.027239    0.040380     0.320146  0.190015     0.427452   
Fireplaces    -0.019772   -0.045569     0.235755  0.271364     0.396765   
GarageYrBlt    0.000070    0.080187     0.064324 -0.024812     0.518018   
GarageCars     0.016570   -0.040110     0.269729  0.154871     0.600671   
GarageArea     0.017634   -0.098672     0.323663  0.180403     0.562022   
WoodDeckSF    -0.029643   -0.012579     0.077106  0.171698     0.238923   
OpenPorchSF   -0.000477   -0.006100     0.137454  0.084774     0.308819   
EnclosedPorch  0.002889   -0.012037     0.009790 -0.018340    -0.113937   
3SsnPorch     -0.046635   -0.043825     0.062335  0.020423     0.030371   
ScreenPorch    0.001330   -0.026030     0.037684  0.043160     0.064886   
PoolArea       0.057044    0.008283     0.180868  0.077672     0.065166   
MiscVal       -0.006242   -0.007683     0.001168  0.038068    -0.031406   
MoSold         0.021172   -0.013585     0.010158  0.001205     0.070815   
YrSold         0.000712   -0.021407     0.006768 -0.014261    -0.027347   
SalePrice     -0.021917   -0.084284     0.334901  0.263843     0.790982   

               OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                0.012609  -0.012713     -0.021998   -0.050199   -0.005024   
MSSubClass       -0.059316   0.027850      0.040581    0.022895   -0.069836   
LotFrontage      -0.052820   0.117598      0.082746    0.179283    0.215828   
LotArea          -0.005636   0.014228      0.013788    0.103960    0.214103   
OverallQual      -0.091932   0.572323      0.550684    0.410238    0.239666   
OverallCond       1.000000  -0.375983      0.073741   -0.127788   -0.046231   
YearBuilt        -0.375983   1.000000      0.592855    0.314745    0.249503   
YearRemodAdd      0.073741   0.592855      1.000000    0.179186    0.128451   
MasVnrArea       -0.127788   0.314745      0.179186    1.000000    0.263582   
BsmtFinSF1       -0.046231   0.249503      0.128451    0.263582    1.000000   
BsmtFinSF2        0.040229  -0.049107     -0.067759   -0.072302   -0.050117   
BsmtUnfSF        -0.136841   0.149040      0.181133    0.114184   -0.495251   
TotalBsmtSF      -0.171098   0.391452      0.291066    0.362452    0

The P-value is the probability value that the correlation between these two variables is statistically significant. Normally, we choose a significance level of 0.05, which means that we are 95% confident that the correlation between the variables is significant.

By convention:

- p-value  <  0.001: strong evidence that the correlation is significant
- p-value  <  0.05: moderate evidence that the correlation is significant
- p-value  <  0.1: weak evidence that the correlation is significant
- p-value  >  0.1: no evidence that the correlation is significant

In [23]:
def get_num_obj(dataset):
    
    """
    This function takes a dataset and returns two dataframes from it: one with numerical data and another with object data.
    """
    
    # Get data types:
    data_type = [dataset[i].dtype for i in dataset.columns]

    # Create dataframe with features and their types
    df_dtypes = pd.DataFrame()
    df_dtypes['Feature'] = dataset.columns
    df_dtypes['Data_type'] = data_type

    # Create two different dataframes into numerical and object
    df_num = df_dtypes.loc[(df_dtypes['Data_type'] != 'object')]
    df_obj = df_dtypes.loc[(df_dtypes['Data_type'] == 'object')]
    
    return df_num, df_obj

In [24]:
train_num, train_obj = get_num_obj(train)

Let's start with numerical values:

In [25]:
features = []
df_Pcorr = []
df_Pval = []
# loop over numerical features to get the pearson correlation and the p value for each:
for i,row in train_num.iterrows():
    feature = row['Feature']
    pearson_coef, p_value = stats.pearsonr(train[feature], train['SalePrice'])
    features.append(feature)
    df_Pcorr.append(pearson_coef)
    df_Pval.append(p_value)

In [26]:
# Create dataframe from arrays
df_corr = pd.DataFrame(features)

# Add column names
df_corr.columns = ['Feature']

# Add other columns
df_corr['PearsonCorr'] = df_Pcorr
df_corr['Pvalue'] = df_Pval

# Drop the target row and sort values
df_corr.drop(df_corr.loc[df_corr['Feature']==target].index, inplace=True)
df_corr.sort_values('Pvalue', ascending=True, inplace=True)
#color = sns.light_palette("pink", as_cmap=True)
#colored_df = df_corr.style.background_gradient(cmap=color)
df_corr

Feature  PearsonCorr         Pvalue
4     OverallQual     0.790982  2.185675e-313
16      GrLivArea     0.708624  4.518034e-223
26     GarageCars     0.640409  2.498644e-169
27     GarageArea     0.623431  5.265038e-158
12    TotalBsmtSF     0.613581  9.484229e-152
13       1stFlrSF     0.605852  5.394711e-147
19       FullBath     0.560664  1.236470e-121
23   TotRmsAbvGrd     0.533723  2.772281e-108
6       YearBuilt     0.522897  2.990229e-103
7    YearRemodAdd     0.507101   3.164948e-96
8      MasVnrArea     0.475241   3.924712e-83
25    GarageYrBlt     0.470177   3.550785e-81
24     Fireplaces     0.466929   6.141487e-80
9      BsmtFinSF1     0.386420   3.394110e-53
2     LotFrontage     0.334901   1.346651e-39
28     WoodDeckSF     0.324413   3.972217e-37
14       2ndFlrSF     0.319334   5.764335e-36
29    OpenPorchSF     0.315856   3.493374e-35
20       HalfBath     0.284108   1.650473e-28
3         LotArea     0.263843   1.123139e-24
17   BsmtFullBath     0.227122   1.550344e-18
11      BsmtUnfSF     0.214479   1.182976e-16
21   BedroomAbvGr     0.168213   9.927497e-11
22   KitchenAbvGr    -0.135907   1.860426e-07
30  EnclosedPorch    -0.128578   8.255770e-07
32    ScreenPorch     0.111447   1.972140e-05
33       PoolArea     0.092404   4.073490e-04
1      MSSubClass    -0.084284   1.266472e-03
5     OverallCond    -0.077856   2.912351e-03
35         MoSold     0.046432   7.612758e-02
31      3SsnPorch     0.044584   8.858170e-02
36         YrSold    -0.028923   2.694132e-01
15   LowQualFinSF    -0.025606   3.282073e-01
0              Id    -0.021917   4.026938e-01
34        MiscVal    -0.021190   4.184863e-01
18   BsmtHalfBath    -0.016844   5.201536e-01
10     BsmtFinSF2    -0.011378   6.639987e-01

In [27]:
interest_feature = [row['Feature'] for i,row in df_corr.iterrows() if abs(row['PearsonCorr']) > 0.5 and row['Pvalue']<0.001]
interest_feature

['OverallQual',
 'GrLivArea',
 'GarageCars',
 'GarageArea',
 'TotalBsmtSF',
 '1stFlrSF',
 'FullBath',
 'TotRmsAbvGrd',
 'YearBuilt',
 'YearRemodAdd']

Let's create another function for this:

In [28]:
def get_corr_features(num_features, dataset, target):
    
    """
    This function calculates the Pearson Correlation Coefficient and the p value of certain numerical features (num_features)
    of a dataset with a specific target. It creates a dataframe with these values and returns it.
    It also returns a list of features that have a Pearson Correlation higher than 0.5 and a p value lower than 0.001,
    ordered by "priority".
    """
    
    features = []
    df_Pcorr = []
    df_Pval = []
    # loop over numerical features to get these stats:
    for i,row in num_features.iterrows():
        feature = row['Feature']
        pearson_coef, p_value = stats.pearsonr(dataset[feature], dataset[target])
        features.append(feature)
        df_Pcorr.append(pearson_coef)
        df_Pval.append(p_value)
    
    # Create dataframe from arrays
    df_corr = pd.DataFrame(features)
    # Add column names
    df_corr.columns = ['Feature']
    # Add other columns
    df_corr['PearsonCorr'] = df_Pcorr
    df_corr['Pvalue'] = df_Pval

    # Drop the target row and sort values
    df_corr.drop(df_corr.loc[df_corr['Feature']==target].index, inplace=True)
    df_corr.sort_values('Pvalue', ascending=True, inplace=True)
    
    # List of features that have a Pearson Correlation higher than 0.5 and a p value lower than 0.001,  ordered by "priority"
    interest_feature = [row['Feature'] for i,row in df_corr.iterrows() if abs(row['PearsonCorr']) > 0.5 and row['Pvalue']<0.001]
    
    print('Features with a Pearson correlation > 0.5 and p value < 0.001:', interest_feature)
    
    return df_corr, interest_feature

In [29]:
train_corr_num, corr_features_num = get_corr_features(train_num, train, target='SalePrice')

Features with a Pearson correlation > 0.5 and p value < 0.001: ['OverallQual', 'GrLivArea', 'GarageCars', 'GarageArea', 'TotalBsmtSF', '1stFlrSF', 'FullBath', 'TotRmsAbvGrd', 'YearBuilt', 'YearRemodAdd']


In [30]:
train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
0        1          60       RL         65.0     8450   Pave      Reg   
1        2          20       RL         80.0     9600   Pave      Reg   
2        3          60       RL         68.0    11250   Pave      IR1   
3        4          70       RL         60.0     9550   Pave      IR1   
4        5          60       RL         84.0    14260   Pave      IR1   
...    ...         ...      ...          ...      ...    ...      ...   
1455  1456          60       RL         62.0     7917   Pave      Reg   
1456  1457          20       RL         85.0    13175   Pave      Reg   
1457  1458          70       RL         66.0     9042   Pave      Reg   
1458  1459          20       RL         68.0     9717   Pave      Reg   
1459  1460          20       RL         75.0     9937   Pave      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   
4          350.0        Gd        TA      PConc       Gd       TA   
..

It's the turn of object features. First, we need to check the number of outcomes each feature can have:

In [31]:
for i,row in train_obj.iterrows():
    feature = row['Feature']
    print(feature)
    print(train[feature].value_counts().count())

MSZoning
5
Street
2
LotShape
4
LandContour
4
Utilities
2
LotConfig
5
LandSlope
3
Neighborhood
25
Condition1
9
Condition2
8
BldgType
5
HouseStyle
8
RoofStyle
6
RoofMatl
8
Exterior1st
15
Exterior2nd
16
MasVnrType
4
ExterQual
4
ExterCond
5
Foundation
6
BsmtQual
4
BsmtCond
4
BsmtExposure
4
BsmtFinType1
6
BsmtFinType2
6
Heating
6
HeatingQC
5
CentralAir
2
Electrical
5
KitchenQual
4
Functional
7
FireplaceQu
5
GarageType
6
GarageFinish
3
GarageQual
5
GarageCond
5
PavedDrive
3
SaleType
9
SaleCondition
6


As we can see, features that have the majority of it's results in one outcome, are not usefull to discern patterns. For example, "Street" has 1454 results for "Pave" and 6 results for Grvl, and "Utilities", has 1459 results for "AllPub" and 1 for NoSeWa.

To start at some point, we can filter out those features whose most frequent outcome has more than 75% of the results:

In [32]:
train_descr = train.describe(include='O').T
train_descr

count unique      top  freq
MSZoning       1460      5       RL  1151
Street         1460      2     Pave  1454
LotShape       1460      4      Reg   925
LandContour    1460      4      Lvl  1311
Utilities      1460      2   AllPub  1459
LotConfig      1460      5   Inside  1052
LandSlope      1460      3      Gtl  1382
Neighborhood   1460     25    NAmes   225
Condition1     1460      9     Norm  1260
Condition2     1460      8     Norm  1445
BldgType       1460      5     1Fam  1220
HouseStyle     1460      8   1Story   726
RoofStyle      1460      6    Gable  1141
RoofMatl       1460      8  CompShg  1434
Exterior1st    1460     15  VinylSd   515
Exterior2nd    1460     16  VinylSd   504
MasVnrType     1460      4     None   872
ExterQual      1460      4       TA   906
ExterCond      1460      5       TA  1282
Foundation     1460      6    PConc   647
BsmtQual       1460      4       TA   686
BsmtCond       1460      4       TA  1348
BsmtExposure   1460      4       No   991
BsmtFinType1   1460      6      Unf   467
BsmtFinType2   1460      6      Unf  1294
Heating        1460      6     GasA  1428
HeatingQC      1460      5       Ex   741
CentralAir     1460      2        Y  1365
Electrical     1460      5    SBrkr  1335
KitchenQual    1460      4       TA   735
Functional     1460      7      Typ  1360
FireplaceQu    1460      5       Gd  1070
GarageType     1460      6   Attchd   951
GarageFinish   1460      3      Unf   686
GarageQual     1460      5       TA  1392
GarageCond     1460      5       TA  1407
PavedDrive     1460      3        Y  1340
SaleType       1460      9       WD  1267
SaleCondition  1460      6   Normal  1198

In [33]:
obj_consider = [i for i,row in train_descr.iterrows() if row['freq'] < train.shape[0]*0.75]
obj_consider

['LotShape',
 'LotConfig',
 'Neighborhood',
 'HouseStyle',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'Foundation',
 'BsmtQual',
 'BsmtExposure',
 'BsmtFinType1',
 'HeatingQC',
 'KitchenQual',
 'FireplaceQu',
 'GarageType',
 'GarageFinish']

We will use indicator variables so we can use categorical variables for regression analysis in the later modules.

An indicator variable (or dummy variable) is a numerical variable used to label categories. They are called 'dummies' because the numbers themselves don't have inherent meaning.

In [34]:
df_train_obj = train.copy()

all_features_corr = train.describe(include='all').T
all_features = [i for i, row in all_features_corr.iterrows()]

for feature in all_features:
#for feature,row in train_descr.iterrows():
#for feature in obj_consider:

    if feature in obj_consider:
    # get dummies
        dummy_var = pd.get_dummies(train[feature])

        # Number of outcomes for feature
        #num_outcomes = train[feature].value_counts().count()
        #for x in range(num_outcomes):

        for column in dummy_var:
            #print(test[column].name)
            # Rename variable
            dummy_var.rename(columns={dummy_var[column].name:feature+'-'+dummy_var[column].name}, inplace=True)
    
        # Concat dataset with dummy 
        df_train_obj = pd.concat([df_train_obj, dummy_var], axis=1)
        
    # drop original column from dataset
    df_train_obj.drop(feature, axis = 1, inplace=True)
    
df_train_obj

LotShape-IR1  LotShape-IR2  LotShape-IR3  LotShape-Reg  \
0                0             0             0             1   
1                0             0             0             1   
2                1             0             0             0   
3                1             0             0             0   
4                1             0             0             0   
...            ...           ...           ...           ...   
1455             0             0             0             1   
1456             0             0             0             1   
1457             0             0             0             1   
1458             0             0             0             1   
1459             0             0             0             1   

      LotConfig-Corner  LotConfig-CulDSac  LotConfig-FR2  LotConfig-FR3  \
0                    0                  0              0              0   
1                    0                  0              1              0   
2                    0                  0              0              0   
3                    1                  0              0              0   
4                    0                  0              1              0   
...                ...                ...            ...            ...   
1455                 0                  0              0              0   
1456                 0                  0              0              0   
1457                 0                  0              0              0   
1458                 0                  0              0              0   
1459                 0                  0              0              0   

      LotConfig-Inside  Neighborhood-Blmngtn  Neighborhood-Blueste  \
0                    1                     0                     0   
1                    0                     0                     0   
2                    1                     0                     0   
3                    0                     0                     0   
4                    0                     0                     0   
...                ...                   ...                   ...   
1455                 1                     0                     0   
1456                 1                     0                     0   
1457                 1                     0                     0   
1458                 1                     0                     0   
1459                 1                     0                     0   

      Neighborhood-BrDale  Neighborhood-BrkSide  Neighborhood-ClearCr  \
0                       0                     0                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   
...                   ...                   ...                   ...   
1455                    0                     0                     0   
1456                    0                     0                     0   
1457                    0                     0                     0   
1458                    0                     0                     0   
1459                    0                     0                     0   

      Neighborhood-CollgCr  Neighborhood-Crawfor  Neighborhood-Edwards  \
0                        1                     0                     0   
1                        0                     0                     0   
2                        1                     0                     0   
3                        0                     1                     0   
4                        0                     0                     0   
...                    ...                   ...                   ...   
1455                     0                     0                     0   
1456                     0

Trying to obtain the Pearson correlation with one-hot encoding for each feature could end up with a matrix that is too complex to extract eny information from it.

We’re looking for a measure of association between two categorical features. Cramér’s V is based on a nominal variation of Pearson’s Chi-Square Test, and comes built-in with some great benefits:

- Similarly to correlation, the output is in the range of [0,1], where 0 means no association and 1 is full association.
- Like correlation, Cramer’s V is symmetrical — it is insensitive to swapping x and y

In [35]:
def cramers_v(x, y):
    """
    Fuction extracted from https://towardsdatascience.com/the-search-for-categorical-correlation-a1cf7f1888c9
    WARNING: This function works specifically to calculate the association between two categorical features (no mix
    with numerical!!)
    """
    confusion_matrix = pd.crosstab(x,y)
    chi2 = stats.chi2_contingency(confusion_matrix)[0]
    n = confusion_matrix.sum().sum()
    phi2 = chi2/n
    r,k = confusion_matrix.shape
    phi2corr = max(0, phi2-((k-1)*(r-1))/(n-1))
    rcorr = r-((r-1)**2)/(n-1)
    kcorr = k-((k-1)**2)/(n-1)
    return np.sqrt(phi2corr/min((kcorr-1),(rcorr-1)))

In [36]:
features = []
cramersv = []
target = 'SalePrice'
# loop over object features to get the stats:
for i,row in train_obj.iterrows():
    feature = row['Feature']
    corr = cramers_v(train[feature], train[target])
    features.append(feature)
    cramersv.append(corr)

# Create dataframe from arrays
df_cram = pd.DataFrame(features)
# Add column names
df_cram.columns = ['Feature']
# Add other columns
df_cram['Cramers_v'] = cramersv
df_cram.sort_values('Cramers_v', ascending=False, inplace=True)
df_cram

Feature  Cramers_v
17      ExterQual   0.444187
29    KitchenQual   0.434299
1          Street   0.393452
20       BsmtQual   0.392090
33   GarageFinish   0.365846
25        Heating   0.349594
27     CentralAir   0.335682
21       BsmtCond   0.333478
2        LotShape   0.324009
18      ExterCond   0.305351
38  SaleCondition   0.296247
0        MSZoning   0.292441
22   BsmtExposure   0.280531
16     MasVnrType   0.264636
37       SaleType   0.262459
19     Foundation   0.221499
31    FireplaceQu   0.208670
7    Neighborhood   0.170330
6       LandSlope   0.148079
5       LotConfig   0.144835
28     Electrical   0.141736
3     LandContour   0.139429
9      Condition2   0.115394
34     GarageQual   0.088966
32     GarageType   0.054944
30     Functional   0.000000
14    Exterior1st   0.000000
36     PavedDrive   0.000000
35     GarageCond   0.000000
15    Exterior2nd   0.000000
4       Utilities   0.000000
8      Condition1   0.000000
26      HeatingQC   0.000000
13       RoofMatl   0.000000
24   BsmtFinType2   0.000000
23   BsmtFinType1   0.000000
11     HouseStyle   0.000000
10       BldgType   0.000000
12      RoofStyle   0.000000

These are the categorical features we could choose, but the numbers are quite poor. The reason for this is that the Cramér’s V is a method that works comparing categoricla features alone.

In [37]:
corr_features_obj = [row['Feature'] for i,row in df_cram.iterrows() if row['Cramers_v'] > 0.35]
corr_features_obj

['ExterQual', 'KitchenQual', 'Street', 'BsmtQual', 'GarageFinish']

The Correlation Ratio would be much more appropriate for our endeavour, because it allows the mix of categorical and numerical variables. It is defined as the weighted variance of the mean of each category divided by the variance of all samples; it answers the following question: Given a continuous number, how well can you know to which category it belongs to? The output is too on the range of [0,1].

In [38]:
def correlation_ratio(categories, measurements):
    """
    Fuction extracted from https://towardsdatascience.com/the-search-for-categorical-correlation-a1cf7f1888c9
    """
    fcat, _ = pd.factorize(categories)
    cat_num = np.max(fcat)+1
    y_avg_array = np.zeros(cat_num)
    n_array = np.zeros(cat_num)
    for i in range(0,cat_num):
        cat_measures = measurements[np.argwhere(fcat == i).flatten()]
        n_array[i] = len(cat_measures)
        y_avg_array[i] = np.average(cat_measures)
    y_total_avg = np.sum(np.multiply(y_avg_array,n_array))/np.sum(n_array)
    numerator = np.sum(np.multiply(n_array,np.power(np.subtract(y_avg_array,y_total_avg),2)))
    denominator = np.sum(np.power(np.subtract(measurements,y_total_avg),2))
    if numerator == 0:
        eta = 0.0
    else:
        eta = np.sqrt(numerator/denominator)
    return eta

In [39]:
corr_ratio = []
features = []

for i,row in train_obj.iterrows():
    feature_obj = row['Feature']
    category = train[feature_obj]
    measure = train['SalePrice']
    eta = correlation_ratio(category, measure)
    
    features.append(feature_obj)
    corr_ratio.append(eta)
    
# Create dataframe from arrays
df_corr_ratio = pd.DataFrame(features)
# Add column names
df_corr_ratio.columns = ['Feature']
# Add other columns
df_corr_ratio['Correlation_Ratio'] = corr_ratio
df_corr_ratio.sort_values('Correlation_Ratio', ascending=False, inplace=True)
df_corr_ratio

Feature  Correlation_Ratio
7    Neighborhood           0.738630
17      ExterQual           0.690933
20       BsmtQual           0.678460
29    KitchenQual           0.675721
33   GarageFinish           0.542375
19     Foundation           0.506328
26      HeatingQC           0.442154
23   BsmtFinType1           0.441780
16     MasVnrType           0.428108
32     GarageType           0.411596
15    Exterior2nd           0.392211
14    Exterior1st           0.390862
37       SaleType           0.370523
22   BsmtExposure           0.369163
38  SaleCondition           0.368100
0        MSZoning           0.327963
31    FireplaceQu           0.318739
11     HouseStyle           0.293790
2        LotShape           0.276362
27     CentralAir           0.251328
28     Electrical           0.244235
12      RoofStyle           0.240201
36     PavedDrive           0.233537
10       BldgType           0.185833
8      Condition1           0.180640
13       RoofMatl           0.177237
21       BsmtCond           0.166229
3     LandContour           0.160605
18      ExterCond           0.153680
34     GarageQual           0.150741
35     GarageCond           0.148509
5       LotConfig           0.144981
30     Functional           0.128376
25        Heating           0.120155
9      Condition2           0.099495
24   BsmtFinType2           0.090287
6       LandSlope           0.051784
1          Street           0.041036
4       Utilities           0.014314

In [40]:
corr_features_obj = [row['Feature'] for i,row in df_corr_ratio.iterrows() if row['Correlation_Ratio'] > 0.5]
corr_features_obj

['Neighborhood',
 'ExterQual',
 'BsmtQual',
 'KitchenQual',
 'GarageFinish',
 'Foundation']

From these results we can see that the previous method could give an idea of the association but some of the possible features weren't actually a good option, like "Street".\
This is hence the list of attributes we will concentrate on for the following analysis:

In [41]:
corr_features = corr_features_num + corr_features_obj
corr_features

['OverallQual',
 'GrLivArea',
 'GarageCars',
 'GarageArea',
 'TotalBsmtSF',
 '1stFlrSF',
 'FullBath',
 'TotRmsAbvGrd',
 'YearBuilt',
 'YearRemodAdd',
 'Neighborhood',
 'ExterQual',
 'BsmtQual',
 'KitchenQual',
 'GarageFinish',
 'Foundation']